provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,218 @@
+		fmt.Sprintf("%sArn", naming.ToSdkName(typeName)),
+	}
+	// These are the most common properties that are stable outputs
+	// and are used to determine if an output is stable or not
+	topLevelStableOutputs := slices.Concat([]string{
+		"name",
+		"id",
+		"arn",
+	}, stableOutputsNameOnly)
+
+	// we can't handle arrays because we don't know which item in the array
+	// the value corresponds to
+	if isArrayProperty(propName) {
+		return false
+	}
+
+	// e.g. aws-native:s3:StorageLens has a storageLensConfiguration/StorageLensArn readOnly property
+	if strings.Contains(propName, "/") {
+		parts := strings.Split(propName, "/")
+		name := parts[len(parts)-1]
+		// If the property is a nested property then only consider it stable if
+		// the property contains the resource type name. There are a lot of cases where
+		// an object has a property called `id` or `arn` that is not  the resource id or arn
+		return slices.Contains(stableOutputsNameOnly, naming.ToSdkName(name))
+	}
+	return slices.Contains(topLevelStableOutputs, naming.ToSdkName(propName))
+}
+
+func isReadOnly(readOnly []string, key string) bool {
+	for _, prop := range readOnly {
+		if naming.ToSdkName(prop) == key {
+			return true
+		}
+	}
+	return false
+}
+
+func filterReadOnly(readOnly []string, key string) []string {
+	result := []string{}
+	for _, prop := range readOnly {
+		if strings.HasPrefix(prop, key+"/") {
+
+			result = append(result, strings.TrimPrefix(prop, key+"/"))
+		}
+	}
+	return result
+}
+
+func isArrayProperty(prop string) bool {
+	return strings.Contains(prop, "/*/")
+}
+
+func arrayReadOnly(readOnly []string) []string {
+	var result []string
+	for _, prop := range readOnly {
+		result = append(result, strings.TrimPrefix(prop, "*/"))
+	}
+	return result
+}
+
+// previewOutputValue exists to recurse through nested objects and populate computed values
+// There are cases where the input and output types are the same, but some properties of the type
+// are only output values.
+func previewOutputValue(
+	inputValue resource.PropertyValue,
+	types map[string]metadata.CloudAPIType,
+	prop *schema.TypeSpec,
+	readOnly []string,
+) resource.PropertyValue {
+	switch {
+	case inputValue.IsArray() && (prop.Type == "array" || prop.Items != nil):
+		var items []resource.PropertyValue
+		for _, item := range inputValue.ArrayValue() {
+			items = append(items, previewOutputValue(item, types, prop.Items, arrayReadOnly(readOnly)))
+		}
+		return resource.NewArrayProperty(items)
+	case inputValue.IsObject() && strings.HasPrefix(prop.Ref, "#/types/"):
+		typeName := strings.TrimPrefix(prop.Ref, "#/types/")
+		if t, ok := types[typeName]; ok {
+			v := previewResourceOutputs(inputValue.ObjectValue(), types, t.Properties, readOnly)
+			return resource.NewObjectProperty(v)
+		}
+	case inputValue.IsObject() && prop.AdditionalProperties != nil:
+		inputObject := inputValue.ObjectValue()
+		result := resource.PropertyMap{}
+		for name, value := range inputObject {
+			p := value
```

Comment:
If I insert a `panic("!")` here and run tests, this line does not get hit. This could be interesting to test out since I think recurring into the map element would need some kind of transformation on the `readOnly` parameter maybe? or if it's not needed it's good to verify.

Reply:
Added some tests and also added some additional logic related to processing `additionalProperties`.